packages/@expo/cli/src/start/server/middleware/inspector/JsInspector.ts

```diff
@@ -15,6 +19,21 @@ export interface MetroInspectorProxyApp {
 
 let openingBrowserInstance: LaunchBrowserInstance | null = null;
 
+const IS_WSL = require('is-wsl') && !require('is-docker')();
+
+function createBrowser() {
```

Comment:
I moved `createBrowser` to **JSInspector.ts** because `createBrowser` was only being used in **LaunchBrowser.ts**. So I could have exported `createBrowser` from **LaunchBrowser.ts** or just move it to **JSInspector.ts**. The former made more sense to me. Nevertheless, if the latter makes more sense I can adjust the changes accordingly.
And I did all of that just to to inject a [`LaunchBrowserImpl`](https://github.com/kapobajza/expo/blob/531d7ff72542b2bc5482d935cd0ff6835937b7f2/packages/@expo/cli/src/start/server/middleware/inspector/LaunchBrowser.ts#L14) instance into the `launchBrowserAsync` function and make it testable. Yeah, sure, I could've used `jest.mock` instead, but I am not fond of `jest.mock` and I find this to be the safer way. But if using `jest.mock` suits this project more, I could also do that.

Reply:
yes, i think exporting `createBrowser` makes more sense and also helps **LaunchBrowser.ts** isolated.

Our test target would then change to `createBrowser` and we coud also move the logic here into the `createBrowser`.
https://github.com/expo/expo/blob/95eaac94793d6a8bdc9a05b341da1899c16b5beb/packages/%40expo/cli/src/start/server/middleware/inspector/LaunchBrowser.ts#L33-L35

does that sound good to you?

Reply:
I've understood the first part and I'll export the `createBrowser` function from **LaunchBrowser.ts**.

But I am not entirely sure what you mean by the second part: to move the browser is supported check to the `createBrowser` function. You mean you want to check if the browser is supported, before returning the `LaunchBrowserImpl` instance from `createBrowser`? Something like:

```
export async function createBrowserAsync(
  platform: NodeJS.Platform,
  supportedBrowsers: LaunchBrowserTypes[]
) {
  let browserImpl: LaunchBrowserImpl | undefined = undefined;

  if (platform === 'darwin') {
    browserImpl = new LaunchBrowserImplMacOS();
  }
  if (platform === 'win32' || IS_WSL) {
    browserImpl = new LaunchBrowserImplWindows();
  }
  if (platform === 'linux') {
    browserImpl = new LaunchBrowserImplLinux();
  }

  if (browserImpl === undefined) {
    throw new Error('[LaunchBrowser] Unsupported host platform');
  }

  let supportedBrowsersCount = 0;

  for (const browserType of supportedBrowsers) {
    const isSupported = await browserImpl.isSupportedBrowser(browserType);
    if (isSupported) {
      supportedBrowsersCount++;
    }
  }

  if (supportedBrowsersCount === 0) {
    throw new Error(
      `[LaunchBrowser] Unable to find a browser on the host to open the inspector. Supported browsers: ${supportedBrowsers.join(
        ', '
      )}`
    );
  }

  return browserImpl;
}
```

If this is what you meant, then I will still have to loop over the `supportedBrowsers` array inside of  `LaunchBrowser`: https://github.com/expo/expo/blob/95eaac94793d6a8bdc9a05b341da1899c16b5beb/packages/%40expo/cli/src/start/server/middleware/inspector/LaunchBrowser.ts#L33-L38

In order to call `browser.launchAsync` with the `browserType` argument.

I have a feeling that I misunderstood you.

Reply:
sorry i didn't describe it clearly. let's show the code directly: https://github.com/expo/expo/compare/%40kudo/pr-25109

if there's still any unclear part, please feel free to let me know.

Reply:
It is clear now. Thank you for the clarification and your code.
I rebased to your branch.